x-pack/plugins/discover_log_explorer/public/state_machines/log_explorer_profile/src/url_state_storage_service.ts

```diff
@@ -78,3 +105,147 @@ const extractDatasetSelectionFromIndex = ({
 
   return datasetSelection;
 };
+
+export const subscribeControlGroup =
+  ({
+    stateContainer,
+  }: LogExplorerProfileUrlStateDependencies): InvokeCreator<
+    LogExplorerProfileContext,
+    LogExplorerProfileEvent
+  > =>
+  (context) =>
+  (send) => {
+    if (!context.controlGroupAPI) return;
+
+    const filtersSubscription = context.controlGroupAPI.onFiltersPublished$.subscribe(
+      (newFilters) => {
+        stateContainer.internalState.transitions.setCustomFilters(newFilters);
+        stateContainer.actions.fetchData();
+      }
+    );
+
+    // Keeps our state in sync with the url changes and makes sure it adheres to correct schema
+    const urlSubscription = stateContainer.stateStorage
+      .change$<ControlPanels>(CONTROL_PANELS_URL_KEY)
+      .subscribe((controlPanels) => {
+        if (!deepEqual(controlPanels, context.controlPanels)) {
+          send({ type: 'UPDATE_CONTROL_PANELS', controlPanels });
+        }
+      });
+
+    // Keeps the url in sync with the controls state after change
+    const inputSubscription = context.controlGroupAPI.getInput$().subscribe(({ panels }) => {
+      if (!deepEqual(panels, context.controlPanels)) {
+        send({ type: 'UPDATE_CONTROL_PANELS', controlPanels: panels });
+      }
+    });
+
+    return () => {
+      filtersSubscription.unsubscribe();
+      urlSubscription.unsubscribe();
+      inputSubscription.unsubscribe();
+    };
+  };
+
+export const updateControlPanels =
+  ({
+    stateContainer,
+  }: LogExplorerProfileUrlStateDependencies): InvokeCreator<
+    LogExplorerProfileContext,
+    LogExplorerProfileEvent
+  > =>
+  async (context, event) => {
+    if (!('controlPanels' in event)) return;
+
+    const newControlPanels = event.controlPanels || context.controlPanels;
+    const controlPanelsWithId = constructControlPanelsWithDataViewId(
+      stateContainer,
+      newControlPanels!
+    );
+
+    context.controlGroupAPI?.updateInput({ panels: controlPanelsWithId });
+
+    return controlPanelsWithId;
+  };
+
+/**
+ * Utils
+ */
+
+const constructControlPanelsWithDataViewId = (
+  stateContainer: DiscoverStateContainer,
+  newControlPanels: ControlPanels
+) => {
+  const dataView = stateContainer.internalState.getState().dataView!;
+
+  const validatedControlPanels = isValidState(newControlPanels)
+    ? newControlPanels
+    : getVisibleControlPanelsConfig(dataView);
+
+  const controlsPanelsWithId = mergeDefaultPanelsWithUrlConfig(dataView, validatedControlPanels!);
+
+  stateContainer.stateStorage.set(
+    CONTROL_PANELS_URL_KEY,
+    cleanControlPanels(controlsPanelsWithId),
+    {
+      replace: true,
+    }
+  );
+
+  return controlsPanelsWithId;
+};
+
+const isValidState = (state: ControlPanels | undefined | null): boolean => {
+  return Object.keys(state ?? {}).length > 0 && ControlPanelRT.is(state);
```

Comment:
**question:** do we need to assert if the object contains any property? Shouldn't be enough to validate it's schema with `ControlPanelRT.is`?